### ISSUE TITLE
[python-package] `params` are lost when copying a `Booster`

### ISSUE THREAD

**Issue by hsorsky** [lightgbm-org/LightGBM#5539]
## Description
A copied `Booster` loses it's `params` attribute

## Reproducible example
```python
import copy

import lightgbm
import pandas as pd
from lightgbm.basic import Booster

train_set = pd.DataFrame(
    {
        "feature1": [1, 2, 3],
        "feature2": [1, 2, 1],
    }
)
booster = Booster(
    params={"num_leaves": 5}, 
    train_set=lightgbm.Dataset(data=train_set),
)
print(booster.params)

copied = copy.deepcopy(booster)
print(copied.params)

```

## Environment info

LightGBM version or commit hash: `3.3.2`

Command(s) you used to install LightGBM

```shell
pip install lightgbm==3.3.2
```

## Additional Comments

Perhaps `Booster.model_from_string` should set it after initialising the object?

**Comment by jameslamb** [comment]
Thanks for this report! I agree that it's desirable for the new Booster's `.params` attribute to match the one it was copied from.

Here's the relevant code run during such a copy.

https://github.com/microsoft/LightGBM/blob/0c0eb2a6a1ba111f0896dd16a579342fa16f7629/python-package/lightgbm/basic.py#L2841-L2847

One approach that might be taken to solve this is to mimic what's done when pickling / unpickling boosters, where every attribute except `handle` (a pointer to a `Booster` object on the C++ side) is preserved via `Booster.__dict__`.

https://github.com/microsoft/LightGBM/blob/0c0eb2a6a1ba111f0896dd16a579342fa16f7629/python-package/lightgbm/basic.py#L2849-L2856

But I'd have to look into it more closely...there are several ways to solve this, I think. @hsorsky are you interested in attempting a pull request? If not, no worries and we're thankful for the report...but wanted to give you the first opportunity.

**Comment by hsorsky** [comment]
Yes, I'm interested in attempting, but I might need some guidance. 

I assume it's preferable to have `Booster.params` be set in `Booster.model_from_string` rather than `Booster__deepcopy__`? 

If `Booster__deepcopy__` suffices, does something like 
```python
    def __deepcopy__(self, _):
        model_str = self.model_to_string(num_iteration=-1)
        booster = Booster(model_str=model_str)
        booster.params = deepcopy(self.params)
        return booster
```
work?

If not, and it would be required to be in `Booster.model_from_string`, does something similar to `_load_pandas_categorical` work but where we'd iterate the lines of `model_str` from `parameters:` to `end of parameters` with custom handling for each param and its expected type (which sounds quite complicated)

**Comment by jmoralez** [comment]
With #5424 merged we can actually get all the params back from the model string, might be something to consider, i.e.
```python
import lightgbm as lgb
import numpy as np

ds = lgb.Dataset(np.random.rand(100, 2))
bst = lgb.train({'num_leaves': 5, 'verbosity': -1}, ds)
bst2 = lgb.Booster(model_str=bst.model_to_string())
print(bst2._get_loaded_param())
```
returns all the parameters. While working on that I remember at one moment I added it somewhere to model_from_string as well and these lines https://github.com/microsoft/LightGBM/blob/0c0eb2a6a1ba111f0896dd16a579342fa16f7629/python-package/lightgbm/engine.py#L267-L268
made it so that the booster that is returned from the train function had all the parameters.

**Comment by jmoralez** [comment]
Ah it was actually this change:
```diff
--- a/python-package/lightgbm/basic.py
+++ b/python-package/lightgbm/basic.py
@@ -2818,12 +2818,13 @@ class Booster:
             self.pandas_categorical = _load_pandas_categorical(file_name=model_file)
             if params:
                 _log_warning('Ignoring params argument, using parameters from model file.')
-            params = self._get_loaded_param()
         elif model_str is not None:
             self.model_from_string(model_str)
         else:
             raise TypeError('Need at least one training dataset or model file or model string '
                             'to create Booster instance')
+        if model_file is not None or model_str is not None:
+            params = self._get_loaded_param()
         self.params = params
 
     def __del__(self) -> None:
```
That'd kind of solve the copy issue. The "problem" is that the copied object actually has more haha.
```python
import copy

import lightgbm as lgb
import numpy as np

ds = lgb.Dataset(np.random.rand(100, 2))
bst = lgb.train({'num_leaves': 5, 'verbosity': -1}, ds)
bst2 = copy.deepcopy(bst)
print(len(bst.params))  # 4
print(len(bst2.params))  # 90
```

**Comment by jameslamb** [comment]
ok yeah nice! At first glance, I like that.

I'm going to be traveling for the next few days and might not be available much. So @hsorsky if you put up a pull request (with tests please 😀 ) I might not be able to look at it until next week, but hopefully @jmoralez will be able to when time permits.

**Comment by jameslamb** [comment]
👋🏻 I'm working on a fix for this right now.

**Comment by afikbar** [comment]
@jameslamb I've just encountered it too, happy to assist with the PR if too busy (:
